hazelcast/include/hazelcast/client/serialization/generic_record.h

```diff
@@ -0,0 +1,1265 @@
+
+class HAZELCAST_API generic_record
+{
+public:
+    /**
+     * Creates a {@link generic_record_builder} allows to create a new object.
+     * This method is a convenience method to get a builder, without creating
+     * the schema/class definition for this type. Here you can see a new object
+     * is constructed from an existing generic_record with its schema
+     * definition:
+     *
+     * <pre>{@code
+     * generic_record new_record = record.new_builder()
+     *      .set_string("name", "bar")
+     *      .set_int32("id", 4)
+     *      .build();
+     * }</pre>
+     * <p>
+     * See {@link generic_record_builder} to create a
+     * generic_record in
+     * {@link generic_record_builder} to create a generic_record in
+     * Compact format with a different schema.
+     *
+     * @return an empty generic record builder with same schema as
+     * this one
+     */
+    generic_record_builder new_builder() const;
+
+    /**
+     * Returned {@link generic_record_builder} can be used to have exact copy
+     * and also just to update a couple of fields. By default, it will copy all
+     * the fields. So instead of following where only the `id` field is updated,
+     * <pre>{@code
+     * generic_record modified_record = record.newBuilder()
+     *         .set_string("name", record.get_string("name"))
+     *         .set_int64("id", 4)
+     *         .set_string("surname", record.get_string("surname"))
+     *         .set_int32("age", record.get_int32("age"))
+     *         .build();
+     * }</pre>
+     * `new_builder_with_clone` used as follows:
+     * <pre>{@code
+     * generic_record modified_record = record.new_builder_with_clone()
+     *         .set_int32("id", 4)
+     *         .build();
+     * }</pre>
+     *
+     * @return a generic record builder with same schema definition as
+     * this one and populated with same values.
+     */
+    generic_record_builder new_builder_with_clone() const;
+
+    /**
+     * @return set of field names of this generic_record
+     */
+    std::unordered_set<std::string> get_field_names() const;
+
+    /**
+     * Returns the kind of the field for the given field name.
+     * <p>
+     * If the field with the given name does not exist,
+     * {@link field_kind#NOT_AVAILABLE} is returned.
+     * <p>
+     * This method can be used to check the existence of a field, which can be
+     * useful when the class is evolved.
+     *
+     * @param field_name name of the field.
+     * @return kind of the field
+     */
+    field_kind get_field_kind(const std::string& field_name) const;
+
+    /**
+     * @param field_name the name of the field
+     * @return the value of the field
+     * @throws exception::hazelcast_serialization if the field name does not
+     * exist in the schema definition or the type of the field does not match
+     * the one in the schema definition.
+     */
+    bool get_boolean(const std::string& field_name) const;
```

Comment:
Should we add get_char method and throw exception from it as it is in Java implementation.

Reply:
I though it but find it completely redundant. In Java side, they have common interface which both support Portable and Compact but for C++ we only support compact serialization so no need to add `char` related functions since it is not supported by Compact.